morpheus/pipeline/pipeline.py

```diff
@@ -146,43 +176,47 @@ def build(self):
 
         self._srf_pipeline = srf.Pipeline()
 
-        def inner_build(builder: srf.Builder):
-            logger.info("====Building Pipeline====")
-
-            # Get the list of stages and source
-            source_and_stages: typing.List[StreamWrapper] = list(self._sources) + list(self._stages)
-
-            # Now loop over stages
-            for s in source_and_stages:
+        def inner_build(builder: srf.Builder, segment_id: str):
+            segment_graph = self._segment_graphs[segment_id]
 
-                if (s.can_build()):
-                    s.build(builder)
+            # This should be a BFS search from each source nodes; but, since we don't have source stage loops
+            # topo_sort provides a resonable approximation.
+            for stage in networkx.topological_sort(segment_graph):
+                if (stage.can_build()):
+                    stage.build(builder)
 
-            if (not all([x.is_built for x in source_and_stages])):
+            if (not all([x.is_built for x in segment_graph.nodes()])):
                 # raise NotImplementedError("Circular pipelines are not yet supported!")
-                logger.warning("Circular pipeline detected! Building with reduced constraints")
+                logger.warning("Cyclic pipeline graph detected! Building with reduced constraints")
 
-                for s in source_and_stages:
+                for stage in segment_graph.nodes():
+                    if (stage.can_build(check_ports=True)):
+                        stage.build()
 
-                    if (s.can_build(check_ports=True)):
-                        s.build()
-
-            if (not all([x.is_built for x in source_and_stages])):
+            if (not all([x.is_built for x in segment_graph.nodes()])):
```

Comment:
Since we are updating this I would remove the brackets which would pass a generator expression into `all` rather than a constructed list.
```suggestion
            if (not all(x.is_built for x in segment_graph.nodes())):
```